typesafeconfig/typesafe_config_test.go

```diff
@@ -0,0 +1,348 @@
+ */
+
+package typesafeconfig
+
+import (
+	"embed"
+	cp "github.com/otiai10/copy"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/suite"
+	"go.uber.org/zap"
+	"os"
+	"testing"
+)
+
+//go:embed test_resources/*
+var testResources embed.FS
+
+type EmbeddedSubConfig struct {
+	SomeOtherStringOption string
+}
+type Config struct {
+	FeatureEnabled    bool
+	NumberOfWidgets   int
+	SomeStringOption  string
+	SomeUnsetValue    string
+	EmbeddedSubConfig EmbeddedSubConfig
+}
+
+type TypesafeConfigTestSuite struct {
+	logicalTestResourcePath string
+	log                     *zap.SugaredLogger
+	suite.Suite
+	VariableThatShouldStartAtFive int
+}
+
+// Make sure that VariableThatShouldStartAtFive is set to five
+// before each test
+func (s *TypesafeConfigTestSuite) SetupSuite() {
+	logger, _ := zap.NewDevelopment()
+	s.log = logger.Sugar()
+
+	dname, err := os.MkdirTemp("", "typesafe-config-test")
+	if err != nil {
+		s.T().Fatal(err)
+	}
+
+	cp.Copy("./test_resources", dname)
+	s.logicalTestResourcePath = dname
+}
+
+func (s *TypesafeConfigTestSuite) TearDownSuite() {
+	os.RemoveAll(s.logicalTestResourcePath)
+}
+
+func (s *TypesafeConfigTestSuite) TestSetValue() {
+	type kvPair struct {
+		key   []string
+		value string
+	}
+	type test struct {
+		name     string
+		kvPairs  []kvPair
+		config   map[string]interface{}
+		expected map[string]interface{}
+	}
+
+	tests := []test{
+		{
+			name: "test that a nested key can be set into a new map",
+			kvPairs: []kvPair{
+				{
+					key:   []string{"foo", "bar", "bam"},
+					value: "baz",
+				},
+				{
+					key:   []string{"foo", "bar", "bop"},
+					value: "wow",
+				},
+			},
+			config: make(map[string]interface{}),
+			expected: map[string]interface{}{
+				"foo": map[string]interface{}{
+					"bar": map[string]interface{}{
+						"bam": "baz",
+						"bop": "wow",
+					},
+				},
+			},
+		},
+		{
+			name: "test that values can be overridden",
+			kvPairs: []kvPair{
+				{
+					key:   []string{"foo", "bar", "bam"},
+					value: "baz",
+				},
+				{
+					key:   []string{"foo", "bar", "bam"},
+					value: "overridden",
+				},
+			},
+			config: make(map[string]interface{}),
+			expected: map[string]interface{}{
+				"foo": map[string]interface{}{
+					"bar": map[string]interface{}{
+						"bam": "overridden",
+					},
+				},
+			},
+		},
+		{
+			name: "test that a key that has a value is overridden by proceeding nested config",
+			kvPairs: []kvPair{
+				{
+					key:   []string{"foo", "bar", "bam"},
+					value: "value1",
+				},
+				{
+					key:   []string{"foo", "bar", "bam", "baz"},
+					value: "some-value",
+				},
+			},
+			config: make(map[string]interface{}),
+			expected: map[string]interface{}{
+				"foo": map[string]interface{}{
+					"bar": map[string]interface{}{
+						"bam": map[string]interface{}{
+							"baz": "some-value",
+						},
+					},
+				},
+			},
+		},
+	}
+
+	for _, tc := range tests {
+		for _, kvPair := range tc.kvPairs {
```

Comment:
use `t.Run(...)` here for table-driven tests

this splits out the pass/fail results for each sub-test
and Goland/Intellij will give you a little green arrow for each sub-test

Reply:
I guess it's `s.T().Run` in a suite